turbopack/crates/turbo-tasks/src/capture_future.rs

```diff
@@ -80,19 +100,46 @@ impl<T, F: Future<Output = T>> Future for CaptureFuture<T, F> {
         EXTRA.with_borrow_mut(|cell| {
             *cell = Some(&mut data as *mut ThreadLocalData);
         });
-        let result = this.future.poll(cx);
+
+        let result =
+            panic::catch_unwind(panic::AssertUnwindSafe(|| this.future.poll(cx))).map_err(|err| {
+                let message = match err.downcast_ref::<&'static str>() {
+                    Some(s) => TurboTasksExecutionErrorMessage::PIISafe(s),
+                    None => match err.downcast_ref::<String>() {
+                        Some(s) => TurboTasksExecutionErrorMessage::NonPIISafe(s.clone()),
+                        None => {
+                            let error_message = err
+                                .downcast_ref::<Box<dyn Error + 'static>>()
+                                .map(|e| e.to_string())
+                                .unwrap_or_else(|| String::from("<unknown panic>"));
+
+                            TurboTasksExecutionErrorMessage::NonPIISafe(error_message)
+                        }
+                    },
+                };
+
+                LAST_ERROR_LOCATION.with_borrow(|loc| TurboTasksPanic {
+                    message,
+                    location: loc.clone(),
+                })
+            });
+
         drop(guard);
         let elapsed = start.elapsed();
         let allocations = start_allocations.until_now();
         *this.duration += elapsed + data.duration;
         *this.allocations += allocations.allocations + data.allocations;
         *this.deallocations += allocations.deallocations + data.deallocations;
         match result {
-            Poll::Ready(r) => {
+            Err(err) => {
+                let memory_usage = this.allocations.saturating_sub(*this.deallocations);
+                Poll::Ready((Err(err), *this.duration, memory_usage))
+            }
+            Ok(Poll::Ready(r)) => {
                 let memory_usage = this.allocations.saturating_sub(*this.deallocations);
-                Poll::Ready((r, *this.duration, memory_usage))
+                Poll::Ready((Ok(r), *this.duration, memory_usage))
             }
-            Poll::Pending => Poll::Pending,
+            Ok(Poll::Pending) => Poll::Pending,
```

Comment:
Not really your fault, but this could be rewritten a little cleaner with the `ready!()` macro: https://doc.rust-lang.org/std/task/macro.ready.html